From 8aa51189cdf9cfec41332da82188531741cbea75 Mon Sep 17 00:00:00 2001
From: Vitaly Kuznetsov <vkuznets@redhat.com>
Date: Mon, 12 May 2014 12:20:32 -0400
Subject: [virt] xen/pvhvm: Initialize xen panic handler for PVHVM guests

Message-id: <1399897232-20381-1-git-send-email-vkuznets@redhat.com>
Patchwork-id: 79675
O-Subject: [RHEL7.1 PATCH BZ839966] xen/pvhvm: Initialize xen panic handler for PVHVM guests
Bugzilla: 839966
RH-Acked-by: Laszlo Ersek <lersek@redhat.com>
RH-Acked-by: Andrew Jones <drjones@redhat.com>
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=839966
Brew build: https://brewweb.devel.redhat.com/taskinfo?taskID=7438427
Test Status: compile-tested, tested over Fedora-20 host with xen-4.3.2
by me

Author: Vaughan Cao <vaughan.cao@oracle.com>
Date: Fri, 16 Aug 2013 16:10:56 +0800

(cherry picked from commit 669b0ae961e87c824233475e987b2d39996d4849)

kernel use callback linked in panic_notifier_list to notice others when panic
happens.
NORET_TYPE void panic(const char * fmt, ...){
    ...
    atomic_notifier_call_chain(&panic_notifier_list, 0, buf);
}
When Xen becomes aware of this, it will call xen_reboot(SHUTDOWN_crash) to
send out an event with reason code - SHUTDOWN_crash.

xen_panic_handler_init() is defined to register on panic_notifier_list but
we only call it in xen_arch_setup which only be called by PV, this patch is
necessary for PVHVM.

Without this patch, setting 'on_crash=coredump-restart' in PVHVM guest config
file won't lead a vmcore to be generate when the guest panics. It can be
reproduced with 'echo c > /proc/sysrq-trigger'.

Signed-off-by: Vaughan Cao <vaughan.cao@oracle.com>
Signed-off-by: Konrad Rzeszutek Wilk <konrad.wilk@oracle.com>
Acked-by: Joe Jin <joe.jin@oracle.com>
Signed-off-by: Vitaly Kuznetsov <vkuznets@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/xen/enlighten.c b/arch/x86/xen/enlighten.c
index 1e3cb02..a9d8079 100644
--- a/arch/x86/xen/enlighten.c
+++ b/arch/x86/xen/enlighten.c
@@ -1711,6 +1711,8 @@ static void __init xen_hvm_guest_init(void)
 
  xen_hvm_init_shared_info();
 
+ xen_panic_handler_init();
+
  if (xen_feature(XENFEAT_hvm_callback_vector))
   xen_have_vector_callback = 1;
  xen_hvm_smp_init();
-- 
1.7.1